### PR TITLE
feat(loki): user definable comparators and operator packages

### PR DESCRIPTION
**Pull request by obeliskos**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/LokiJS-Forge/LokiDB/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Currently, comparisons between unindexed finds, indexed finds, and sorting are utilizing the same 'helper' functions which are not easily 'tuned' for performance.  Only one implementation currently exists for our opererators within 'LokiOps'

Issue Number: N/A

## What is the new behavior?
A named registration map of Comparators now exists for reference throughout LokiDB.  Multiple implementations of our operator packages now exist to allow backwards compatibility or tuned for different purposes or functionality.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: docs/operator_packages.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+## What are the LokiOperatorPackages which LokiDB supports?
+Out of the box, LokiDB implements the following operator packages, which you can choose from (currently per-collection) :
+
+- **"js"** (default) : the default implementation using fastest javascript comparisons. This is an instance of the main **LokiOperatorPackage** class.
+- **"loki"** : implementation recommended for dirty data / mixed datatypes. This is an instance of **LokiAbstractOperator** class.
+- **_comparator_** - you need to instantiate this yourself and register that with a name within the loki constructor (example below).  These are created by instantiating the **ComparatorOperatorPackage** class, passing in a comparator to its constructor.
+
+In practice we have one main implementation of all ops in the (default) "LokiOperatorPackage" class.  Other implementations
```
Comment (Viatorus) [inline docs/operator_packages.md]: A missing sentence at the end.

File: docs/ranged_indexes.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+```
+
+## What other indexes might I want to have an implementation for in the future?
+Honestly, most people probably will not need anything other than our AVL index implementation.
+
+The AVL tree is known for keeping strict balancing of its property's values as documents are inserted, updated and removed.
+
+Other implementations such as the "Red-Black" balanced binary search tree algorithms are less strictly balanced so they may have slightly faster inserts, updates, and removes while potentially having slightly slower lookups/finds. 
+
+## If i implement my own RangedIndex, how might I register and use that IRangedIndex implemenation?
```
Comment (Viatorus) [inline docs/ranged_indexes.md]: If I

File: packages/loki/spec/generic/cloning.spec.ts [hunk 3]
```diff
@@ -51,12 +51,14 @@ describe("cloning behavior", () => {
     });
   });
 
+  /*
```
Comment (Viatorus) [inline packages/loki/spec/generic/cloning.spec.ts]: Comment out tests?

File: packages/loki/spec/generic/dynamic_view.spec.ts [hunk 4]
```diff
@@ -453,54 +453,11 @@ describe("dynamicviews", () => {
...
       expect(results.length).toBe(5);
       for (let idx = 0; idx < results.length - 1; idx++) {
-        expect(LokiOps.$lte(results[idx]["b"], results[idx + 1]["b"]));
+        expect(LokiOperatorPackageMap["js"].$lte(results[idx]["b"], results[idx + 1]["b"]));
       }
 
       // remove dynamic view
       coll.removeDynamicView("dvtest");
-
-      // add basic dv with filter on a and simplesort (with js fallback) on b
```
Comment (Viatorus) [inline packages/loki/spec/generic/dynamic_view.spec.ts]: Dynamic view tests are obsolete?

File: packages/loki/spec/generic/sortingIndexing.spec.ts [hunk 5]
```diff
@@ -294,6 +292,7 @@ describe("sorting and indexing", () => {
       expect(indexVals[13] === "asdf").toEqual(true);
     });
 
+    /*
```
Comment (Viatorus) [inline packages/loki/spec/generic/sortingIndexing.spec.ts]: Comment out?

File: docs/operator_packages.md (hunk 1 above)
Comment (obeliskos) [reply docs/operator_packages.md]: fixed

File: docs/ranged_indexes.md (hunk 2 above)
Comment (obeliskos) [reply docs/ranged_indexes.md]: fixed

File: packages/loki/spec/generic/dynamic_view.spec.ts (hunk 4 above)
Comment (obeliskos) [reply packages/loki/spec/generic/dynamic_view.spec.ts]: yes, that one is... 
simplesort no longer has options for : 
- js fallback : now that we allow you to pass in the sortcomparer
- index intercept : no longer using this obsolete optimization based on binary indices (maybe if we rework data array to contain loki id's we can reintroduce.

File: packages/loki/spec/generic/sortingIndexing.spec.ts (hunk 5 above)
Comment (obeliskos) [reply packages/loki/spec/generic/sortingIndexing.spec.ts]: Added back core logic for that one unit test;  

Before going live with v1.0 I will probably want to reassess date operators either in default "js" operator package or "loki" abstract operator package (which we will make default when upgrading legacy lokijs databases)

**Comment by obeliskos** [comment]
Thanks for comments, seems dates were giving me problems earlier so I commented out and forgot.

Not sure I have best strategy for dates at the moment but this infrastructure we should be able to 'blend' in date functionality as we want to shape it in future as well as support for past behavior (in loki abstract operator package).  I think maybe as part of release planning we can discuss best mix of operator package functionality for both.

**Pull request merged**